### ISSUE TITLE
Error when pushing ECR public registry: `unable to get digest: Got bad response from registry: 400 Bad Request`

### ISSUE THREAD

**Issue by armona** [kreuzwerker/terraform-provider-docker#622]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and docker Provider) Version

OpenTofu v1.7.1
on darwin_arm64
+ provider registry.opentofu.org/hashicorp/aws v5.51.1
+ provider registry.opentofu.org/kreuzwerker/docker v3.0.2

### Affected Resource(s)

* `docker_registry_image`

### Terraform Configuration Files

```hcl
terraform {
  required_version = ">= 1.2"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 5.0"
    }

    docker = {
      source  = "kreuzwerker/docker"
      version = "~> 3.0"
    }
  }
}

variable "image_name" {
  type = string
}

variable "dockerfile" {
  type = string
  default = "Dockerfile"
}

provider "docker" {
  registry_auth {
    address  = "public.ecr.aws"
    username = data.aws_ecrpublic_authorization_token.token.user_name
    password = data.aws_ecrpublic_authorization_token.token.password
  }
}

data "aws_ecrpublic_authorization_token" "token" {}

resource "docker_image" "builder" {
  name = var.image_name

  build {
    dockerfile = var.dockerfile
    context    = path.module

    auth_config {
      host_name = "public.ecr.aws"
      auth      = data.aws_ecrpublic_authorization_token.token.authorization_token
    }
  }
}

resource "docker_registry_image" "this" {
  name          = docker_image.builder.name
  keep_remotely = true
}
```

### Expected Behaviour

The image should be pushed to remote registry, and finish without error.

### Actual Behaviour

The image was pushed successfully, but the terraform apply failed with the following error:
```
╷
│ Error: Unable to create image, image not found: unable to get digest: Got bad response from registry: 400 Bad Request
│
│   with docker_registry_image.default[0],
│   on main.tf line 113, in resource "docker_registry_image" "default":
│  113: resource "docker_registry_image" "default" {

```

### Steps to Reproduce

1. Create a public [ECR registry](https://docs.aws.amazon.com/AmazonECR/latest/public/public-repositories.html) 
2. `terraform apply`

### References

* Basically the same issue as https://github.com/kreuzwerker/terraform-provider-docker/issues/356

**Comment by jjfuentes02** [comment]
@armona I am also experiencing same issue. Any luck figuring this out? Otherwise, will have to handle the building and pushing to ECR without using Terraform :(.

**Comment by armona** [comment]
@jjfuentes02 Unfortunately no, couldn't find a workaround using this provider.

**Comment by skyfenton** [comment]
Having the same issue, repo address and tags are correct, just gives 400 error.

**Comment by mtcderek** [comment]
I'm having the same issue. I've tried everything I can find. The image pushes fine and I can pull it back to my system using the docker_container resource just fine. Everything is great except for the error. I'm trying to use the provider in an upcoming Terraform course, but this might force me to go a different direction.

**Comment by CreativeManix** [comment]
I have same issue, I am using  cdktf with c#. Image created locally, and also uploaded to ECR but still i get following error

```
tf-csharp  ╷
           │ Error: Unable to create image, image not found: unable to get digest: Got bad response from registry: 404 Not Found
           │ 
           │   with docker_registry_image.DockerImagePush (DockerImagePush),
           │   on cdk.tf.json line 105, in resource.docker_registry_image.DockerImagePush (DockerImagePush):
           │  105:       }
           │ 

```

This is a blocker for me to use Terraform, same works with Pulumi.

**Comment by Junkern** [comment]
I have the feeling that this is fixed with https://github.com/kreuzwerker/terraform-provider-docker/pull/690 which was released with the newest version `3.2.0`.
Can you try again?

**Comment by mtcderek** [comment]
I'm getting the same error on `3.2.0`:
```
│ Error: Unable to create image, image not found: unable to get digest: Got bad response from registry: 404 Not Found
│ 
│   with docker_registry_image.this,
│   on main.tf line 53, in resource "docker_registry_image" "this":
│   53: resource "docker_registry_image" "this" {
```
Same behavior:
1. apply works and image is pushed to the registry
2. Error occurs, halting Terraform.
3. the docker_registry_image is never applied to state, but the image is pushed. 

I'm providing authentication information to the provider like so:
```hcl
provider "docker" {
  registry_auth {
    address    = local.ecr_url
    username   = local.ecr_token.user_name
    password   = local.ecr_token.password
  }
}

resource "docker_image" "this" {
  name = "${local.ecr_url}:1.0.1"
  build {
    context     = path.module
    dockerfile  = "${path.module}/Dockerfile"
  }
}

resource "docker_registry_image" "this" {
  name          = docker_image.this.name
}
```

**Comment by fr33z3** [comment]
Exactly the same issue for me. Has anyone found a work around?

**Comment by fr33z3** [comment]
Ok I think I have figured this out. So for anyone who will struggle.
Make sure that `registry_auth.address` is a of a format `<id>.dkr.ecr.<zone>.amazonaws.com`. AWS ECR gives ecr_url which includes image name. So you will have something like `<id>.dkr.ecr.<zone>.amazonaws.com/<image-name>`. So if you use ecr_url make a split `split("/", ecr_url)[0]` to be used in `registry_auth.address`. And use full ecr_url for docker_image.name.

**Comment by Junkern** [comment]
@fr33z3  thanks for the finding! I have added your comment (in a modified way) to the official documentation in https://github.com/kreuzwerker/terraform-provider-docker/pull/736 

In addition I have updated the error handling for public ECR pushes so that it is clearer what might be the cause of failure. Will close this with the merge of the PR
